data-serving/scripts/export-data/full_export.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+# Full data export
+# Depends: aws-cli
+
+set -euo pipefail
+
+COUNTRY_EXPORT_BUCKET="s3://covid-19-country-export"
+FULL_EXPORT_BUCKET="s3://covid-19-data-export"
+
+# Unlike country_export.sh, here FORMAT only takes one value
+# Multiple formats are exported using separate AWS Batch job definitions
+FORMAT="${FORMAT:-csv}"
+FILE="latestdata-$FORMAT.tar"
+FILE_DATE="$(date '+%Y-%m-%d')-$FORMAT.tar"
+
+DIR="$(mktemp -d)"
+echo "Starting full export for format: $FORMAT"
+echo "Made temporary directory $DIR"
+trap "rm -rf $DIR" EXIT  # Cleanup before exit
+
+cp data_dictionary.txt "$DIR"
+cp citation.txt "$DIR"
+
+mkdir "$DIR/country"
+echo "Downloading files from $COUNTRY_EXPORT_BUCKET..."
+aws s3 sync $COUNTRY_EXPORT_BUCKET/$FORMAT/ "$DIR/country"
+echo "Preparing tarball..."
+tar cf $FILE -C "$DIR" .
+echo "Uploading to $FULL_EXPORT_BUCKET..."
+aws s3 cp $FILE $FULL_EXPORT_BUCKET/latest/$FILE
+aws s3 cp $FILE $FULL_EXPORT_BUCKET/archive/$FILE_DATE
```

Comment:
Could this be a Python script instead? These longer-running `aws` operations could benefit from error handling (including a message about what went wrong) and retry logic in case of network issues.

Reply:
If it is about retry logic (which I should use, thanks!) then AWS CLI has retry logic built in which we can use https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-retries.html#cli-usage-retries-configure

All errors are logged to CloudWatch.

If it is about rewriting it in Python, I would prefer not to for this case, as I would have to duplicate the S3 sync functionality in Python, for unclear advantages.

Reply:
It is about retry logic and error handling, to be sure, and also about maintainability and extensibility.
Great that `aws-cli` has retry logic built in; of course we should use that.
OK that errors are logged to CloudWatch. The benefit of doing our own exception handling is that we can add searchable messages, metrics, and provide context to otherwise ambiguous error messages. Unfortunately some error messages don't actually provide much useful diagnostic information.
For this script, adding an `x` to `set -eou` would add at least context around which command failed.

Maintainability and extensibility may seem like unclear advantages, as it's impossible to predict what requirements will change. However, requirements absolutely will change, so using tools that make it easy to adjust clearly wins over using tools that don't. This is what [SOLID programming](https://en.wikipedia.org/wiki/SOLID) is all about, and Bash doesn't really allow for that.

Imagine, further, that a bug is discovered around some behavior managed by Bash. Ideally we'd fix the bug and add an accompanying test to prove the bug is fixed and prevent back-sliding. Bash doesn't really help us in that regard. I like Bash as an adhesive between different system components (e.g. with Docker), but only because I don't know what else to use, and in general think it should not do any kind of "heavy lifting"--some scripting and/or systems language should do that, backed by the confidence given by a suite of automated tests.

Not a blocker, but this kind of approach (untested/untestable system-critical code) does feel a bit like an [unhedged bet](https://www.castsoftware.com/blog/bad-code-isnt-technical-debt-its-an-unhedged-call-option).

Reply:
Agreed with all your points, especially relating to larger pieces of code, for which Bash is unsuitable. I disagree that we have to move to Python now; as this in Batch, it is easier to swap out when requirements change. As it stands, this a short script which is basically a sync tar and copy. We can discuss this in call, if we want to not use Bash going forward in the project due to maintainability and extensibility concerns.

Reply:
Apologies if I implied that we need to move to Python now -- we don't, and it's worth considering and discussing.
And if we're using Batch, which uses Docker, it makes sense to me to simply default to using Python containers. I understand why it was used in Lambda (to limit layers) but I don't think it makes sense going forward. 
For sure, let's talk about it tomorrow.